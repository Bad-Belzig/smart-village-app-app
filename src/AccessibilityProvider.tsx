import React, { createContext, useEffect, useState } from 'react';

import { accessibilityListeners } from './helpers';

const defaultAccessibility = {
  isBoldTextEnabled: false,
  isGrayscaleEnabled: false,
  isInvertColorsEnabled: false,
  isReduceMotionEnabled: false,
  isReduceTransparencyEnabled: false,
  isScreenReaderEnabled: false
};

export const AccessibilityContext = createContext(defaultAccessibility);

export const AccessibilityProvider = ({ children }: { children?: React.ReactNode }) => {
  const [accessibility, setAccessibility] = useState(defaultAccessibility);

  useEffect(() => {
    accessibilityListeners(setAccessibility);
  }, []);

  return (
    <AccessibilityContext.Provider value={accessibility}>{children}</AccessibilityContext.Provider>
  );
};
