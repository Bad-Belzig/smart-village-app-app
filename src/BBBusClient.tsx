import { InMemoryCache } from 'apollo-cache-inmemory';
import ApolloClient from 'apollo-client';
import { createHttpLink } from 'apollo-link-http';

import { bbBusUrlV2 } from './config';

const httpLink = createHttpLink({
  uri: bbBusUrlV2
});

export const BBBusClient = new ApolloClient({
  cache: new InMemoryCache(),
  link: httpLink
});
