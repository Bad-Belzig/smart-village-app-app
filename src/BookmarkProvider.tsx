import React, { createContext, useCallback, useEffect, useState } from 'react';

import {
  BookmarkList,
  BOOKMARKS_STORE_KEY,
  readFromStore,
  toggleBookmark as toggleBookmarkHelper
} from './helpers';

type BookmarkProviderValues = {
  bookmarks?: BookmarkList;
  toggleBookmark: (itemType: string, id: string, suffix?: number | string) => Promise<void>;
};

// if we try to toggle a bookmark, while the component is not wrapped
// inside of a provider simply do nothing.
export const BookmarkContext = createContext<BookmarkProviderValues>({
  toggleBookmark: () => new Promise<void>((resolve) => resolve())
});

export const BookmarkProvider = ({ children }: { children?: React.ReactNode }) => {
  const [bookmarks, setBookmarks] = useState<BookmarkList>();

  const toggleBookmark = useCallback(
    async (itemType: string, id: string, suffix?: number | string) => {
      setBookmarks(await toggleBookmarkHelper(itemType, id, suffix));
    },
    [setBookmarks]
  );

  const loadBookmarks = useCallback(async () => {
    setBookmarks(await readFromStore(BOOKMARKS_STORE_KEY));
  }, [setBookmarks]);

  useEffect(() => {
    loadBookmarks();
  }, [loadBookmarks]);

  return (
    <BookmarkContext.Provider value={{ bookmarks, toggleBookmark }}>
      {children}
    </BookmarkContext.Provider>
  );
};
