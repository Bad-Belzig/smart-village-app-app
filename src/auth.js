import * as SecureStore from 'expo-secure-store';

import { namespace, secrets } from './config';

const ACCESS_TOKEN_EXPIRE_TIME = 'ACCESS_TOKEN_EXPIRE_TIME';

/**
 * Check if a auth token is expired or still valid.
 * The expire time is saved as seconds from 01.01.1970 00:00:00 UTC, so
 * we need to divide Date.now() by 1000, which otherwise would return miliseconds.
 */
const isTokenValid = async () => {
  let accessTokenExpireTime = null;

  // The reason for the problem of staying in SplashScreen that occurs after the application is
  // updated on the Android side is the inability to obtain the token here.
  // For this reason, try/catch is used here and the problem of getting stuck in SplashScreen is solved.
  try {
    accessTokenExpireTime = await SecureStore.getItemAsync(ACCESS_TOKEN_EXPIRE_TIME);
  } catch {
    // Token deleted here so that it can be recreated
    await SecureStore.deleteItemAsync(ACCESS_TOKEN_EXPIRE_TIME);
  }

  if (!accessTokenExpireTime) return false;

  const now = parseInt(Date.now() / 1000, 10); // in seconds from 01.01.1970 00:00:00 UTC
  const expires = parseInt(accessTokenExpireTime, 10); // in seconds from 01.01.1970 00:00:00 UTC

  return now < expires;
};

/**
 * Requests the server to authenticate the mobile app with the main server.
 */
export const auth = async () => {
  // quit if the token is still valid
  if (await isTokenValid()) return;

  // otherwise fetch a new access token and expire time
  const fetchObj = {
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      client_id: secrets[namespace].oAuthClientId,
      client_secret: secrets[namespace].oAuthClientSecret,
      grant_type: secrets[namespace].oAuthGrantType
    })
  };

  const response = await fetch(
    `${secrets[namespace].serverUrl}${secrets[namespace].oAuthTokenEndpoint}`,
    fetchObj
  );
  const json = await response.json();

  await SecureStore.setItemAsync('ACCESS_TOKEN', json.access_token);
  // save the time when the token will expire, calculated from the creation time in seconds
  // added by the expire duration in seconds
  await SecureStore.setItemAsync(ACCESS_TOKEN_EXPIRE_TIME, `${json.created_at + json.expires_in}`);
};
