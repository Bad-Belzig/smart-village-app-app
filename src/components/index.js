export * from './bookmarks';
export * from './infoCard';
export * from './map';
export * from './screens';
export * from './wasteCalendar';
export * from './weather';
export * from './widgets';

export * from './BackToTop';
export * from './Button';
export * from './CardList';
export * from './CardListItem';
export * from './CategoryList';
export * from './CategoryListItem';
export * from './ConnectedImagesCarousel';
export * from './DiagonalGradient';
export * from './DropdownHeader';
export * from './DropdownSelect';
export * from './HeaderLeft';
export * from './HomeSection';
export * from './HtmlView';
export * from './Icon';
export * from './Image';
export * from './ImageSection';
export * from './ImagesCarousel';
export * from './ImagesCarouselItem';
export * from './ImageTextList';
export * from './Label';
export * from './Link';
export * from './ListComponent';
export * from './ListSwitchItem';
export * from './LoadingContainer';
export * from './Logo';
export * from './MapSwitchHeader';
export * from './MediaSection';
export * from './Radiobutton';
export * from './SafeAreaViewFlex';
export * from './ServiceBox';
export * from './SettingsListItem';
export * from './StorySection';
export * from './Switch';
export * from './TabBarIcon';
export * from './Text';
export * from './TextList';
export * from './TextListItem';
export * from './Title';
export * from './ToggleListItem';
export * from './Touchable';
export * from './VersionNumber';
export * from './Wrapper';
