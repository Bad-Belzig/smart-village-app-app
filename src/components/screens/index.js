export * from './About';
export * from './EventRecord';
export * from './LunchSection';
export * from './NewsItem';
export * from './OpeningTimesCard';
export * from './OperatingCompany';
export * from './PointOfInterest';
export * from './PriceCard';
export * from './Service';
export * from './Tour';
