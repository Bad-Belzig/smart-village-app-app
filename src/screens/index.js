export * from './AboutScreen';
export * from './BookmarkScreen';
export * from './BookmarkCategoryScreen';
export * from './ConstructionSiteDetailScreen';
export * from './ConstructionSiteOverviewScreen';
export * from './CompanyScreen';
export * from './DataProviderScreen';
export * from './DetailScreen';
export * from './FormScreen';
export * from './HomeScreen';
export * from './HtmlScreen';
export * from './IndexScreen';
export * from './LunchScreen';
export * from './OParl';
export * from './WasteCollectionScreen';
export * from './WasteReminderScreen';
export * from './ServiceScreen';
export * from './SettingsScreen';
export * from './SurveyDetailScreen';
export * from './SurveyOverviewScreen';
export * from './WeatherScreen';
export * from './WebScreen';
