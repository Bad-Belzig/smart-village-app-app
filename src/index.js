import AsyncStorage from '@react-native-async-storage/async-storage';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { persistCache } from 'apollo-cache-persist';
import { ApolloClient } from 'apollo-client';
import { ApolloLink } from 'apollo-link';
import { setContext } from 'apollo-link-context';
import { createHttpLink } from 'apollo-link-http';
import * as ScreenOrientation from 'expo-screen-orientation';
import * as SecureStore from 'expo-secure-store';
import * as SplashScreen from 'expo-splash-screen';
import _isEmpty from 'lodash/isEmpty';
import React, { useEffect, useState } from 'react';
import { ApolloProvider } from 'react-apollo';
import { ActivityIndicator, StatusBar } from 'react-native';
import { SafeAreaProvider } from 'react-native-safe-area-context';

import { auth } from './auth';
import { BookmarkProvider } from './BookmarkProvider';
import { LoadingContainer } from './components';
import { colors, consts, namespace, secrets } from './config';
import { graphqlFetchPolicy, parsedImageAspectRatio, storageHelper } from './helpers';
import { Navigator } from './navigation/Navigator';
import NetInfo from './NetInfo';
import { NetworkProvider } from './NetworkProvider';
import { OrientationProvider } from './OrientationProvider';
import { getQuery, QUERY_TYPES } from './queries';
import { SettingsProvider } from './SettingsProvider';

const { LIST_TYPES } = consts;

const MainAppWithApolloProvider = () => {
  const [loading, setLoading] = useState(true);
  const [isSplashScreenVisible, setIsSplashScreenVisible] = useState(true);
  const [client, setClient] = useState();
  const [initialGlobalSettings, setInitialGlobalSettings] = useState({});
  const [initialListTypesSettings, setInitialListTypesSettings] = useState({});
  const [authRetried, setAuthRetried] = useState(false);
  const [netInfo, setNetInfo] = useState({
    isConnected: null,
    isMainserverUp: null,
    netInfoCounter: 0
  });
  const { isConnected, isMainserverUp, netInfoCounter } = netInfo;

  const setupApolloClient = async () => {
    // https://www.apollographql.com/docs/react/recipes/authentication/#header
    const httpLink = createHttpLink({
      uri: `${secrets[namespace].serverUrl}${secrets[namespace].graphqlEndpoint}`
    });
    const authLink = setContext(async (_, { headers }) => {
      // get the authentication token from local SecureStore if it exists
      const accessToken = await SecureStore.getItemAsync('ACCESS_TOKEN');

      // return the headers to the context so httpLink can read them
      return {
        headers: {
          ...headers,
          authorization: accessToken ? `Bearer ${accessToken}` : ''
        }
      };
    });
    // Note: httpLink is terminating so must be last, while retry & error wrap the links to their right
    //       state & context links should happen before (to the left of) restLink.
    //       https://www.apollographql.com/docs/link/links/rest/#link-order
    // const link = ApolloLink.from([authLink, restLink, errorLink, retryLink, httpLink]);
    const link = ApolloLink.from([authLink, httpLink]);
    const cache = new InMemoryCache();
    const storage = AsyncStorage;

    try {
      // await before instantiating ApolloClient,
      // else queries might run before the cache is persisted
      await persistCache({
        cache,
        storage
      });
    } catch (error) {
      console.error('Error restoring Apollo cache', error);
    }

    // Setup your Apollo Link, and any other Apollo packages here.
    const client = new ApolloClient({
      link,
      cache,

      // From the docs: https://www.apollographql.com/docs/react/essentials/local-state#client-fields-cache
      // If you want to use Apollo Client's @client support to query the cache without using
      // local resolvers, you must pass an empty object into the ApolloClient constructor
      // resolvers option. Without this Apollo Client will not enable its integrated @client
      // support, which means your @client based queries will be passed to the Apollo Client
      // link chain. You can find more details about why this is necessary here
      // (https://github.com/apollographql/apollo-client/pull/4499).
      resolvers: {}
    });

    setClient(client);
  };

  // we can provide an empty array as second argument to the effect hook to avoid activating
  // it on component updates but only for the mounting of the component.
  // this effect depend on no variables, so it is only triggered when the component mounts.
  // if an effect depends on a variable (..., [variable]), it is triggered every time it changes.
  // provide different effects for different contexts.
  //
  // we wait for NetInfo to check for main server reachability, which is made when `isMainserverUp`
  // becomes `true` or `false` and is not `null` anymore. with the `netInfoCounter` that gets
  // updated every time the main server was not checked for reachability, we trigger the the
  // effect again until NetInfo has finished.
  useEffect(() => {
    const updateNetInfo = async () => {
      const updatedNetInfo = await NetInfo.fetch();

      setNetInfo({
        isConnected: updatedNetInfo.isConnected,
        isMainserverUp: updatedNetInfo.isInternetReachable,
        netInfoCounter: updatedNetInfo.isInternetReachable === null && netInfoCounter + 1
      });
    };

    // re-run NetInfo
    isMainserverUp === null && updateNetInfo();

    // setup the apollo client if NetInfo finished and if there is no client existing already
    isMainserverUp !== null && !client && auth(setupApolloClient);
  }, [netInfoCounter]);

  const setupInitialGlobalSettings = async () => {
    const fetchPolicy = graphqlFetchPolicy({ isConnected, isMainserverUp });

    // rehydrate data from the async storage to the global state
    // if there are no general settings yet, add a navigation fallback
    let globalSettings = await storageHelper.globalSettings();

    // if there are no list type settings yet, set the defaults as fallback
    const listTypesSettings = (await storageHelper.listTypesSettings()) || {
      [QUERY_TYPES.NEWS_ITEMS]: LIST_TYPES.TEXT_LIST,
      [QUERY_TYPES.EVENT_RECORDS]: LIST_TYPES.TEXT_LIST,
      [QUERY_TYPES.POINTS_OF_INTEREST_AND_TOURS]: LIST_TYPES.CARD_LIST
    };

    let globalSettingsData;

    try {
      const response = await client.query({
        query: getQuery(QUERY_TYPES.PUBLIC_JSON_FILE),
        variables: { name: 'globalSettings' },
        fetchPolicy
      });

      globalSettingsData = response.data;
    } catch (error) {
      console.warn('error', error);

      if (error.message.includes('Network error')) {
        // try once to authenticate with forcing a fresh token request
        !authRetried && auth(setupApolloClient, true);
        // set flag `true` to prevent endless loops
        setAuthRetried(true);
      }
    }

    let globalSettingsPublicJsonFileContent = [];

    try {
      globalSettingsPublicJsonFileContent = JSON.parse(globalSettingsData?.publicJsonFile?.content);
    } catch (error) {
      console.warn(error, globalSettingsData);
    }

    if (!_isEmpty(globalSettingsPublicJsonFileContent)) {
      globalSettings = globalSettingsPublicJsonFileContent;
      storageHelper.setGlobalSettings(globalSettings);
    }

    setInitialGlobalSettings(globalSettings);
    setInitialListTypesSettings(listTypesSettings);
  };

  // setup global settings if apollo client setup finished
  useEffect(() => {
    client && setupInitialGlobalSettings();
  }, [client]);

  useEffect(() => {
    // this is currently the last point where something was done, so the app startup is done
    initialGlobalSettings && client && setLoading(false);
  }, [initialGlobalSettings]);

  useEffect(() => {
    !loading &&
      SplashScreen.hideAsync().then(() => {
        setIsSplashScreenVisible(false);
        // set orientation to "default", to allow both portrait and landscape
        ScreenOrientation.lockAsync(ScreenOrientation.OrientationLock.DEFAULT);
      });
  }, [loading]);

  if (loading || isSplashScreenVisible) {
    return (
      <LoadingContainer>
        <ActivityIndicator color={colors.accent} />
      </LoadingContainer>
    );
  }

  if (initialGlobalSettings.imageAspectRatio) {
    consts.IMAGE_ASPECT_RATIO = parsedImageAspectRatio(initialGlobalSettings.imageAspectRatio);
  }

  return (
    <ApolloProvider client={client}>
      <SettingsProvider {...{ initialGlobalSettings, initialListTypesSettings }}>
        <StatusBar barStyle="light-content" translucent backgroundColor="transparent" />
        <Navigator />
      </SettingsProvider>
    </ApolloProvider>
  );
};

export const MainApp = () => (
  <NetworkProvider>
    <OrientationProvider>
      <BookmarkProvider>
        <SafeAreaProvider>
          <MainAppWithApolloProvider />
        </SafeAreaProvider>
      </BookmarkProvider>
    </OrientationProvider>
  </NetworkProvider>
);
