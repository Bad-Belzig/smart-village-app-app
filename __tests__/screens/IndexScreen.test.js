import React from 'react';
import renderer from 'react-test-renderer';

import { IndexScreen } from '../../src/screens';

describe('IndexScreen', () => {
  const navigation = { navigate: jest.fn() };

  test.skip('renders correctly', () => {
    const tree = renderer.create(<IndexScreen navigation={navigation} />).toJSON();
    expect(tree).toBeTruthy();
  });

  test.skip('must contain a left header element (for going back)', () => {
    const navigationOptions = IndexScreen.navigationOptions({ navigation });
    const leftHeaderElement = navigationOptions.headerLeft;

    expect(leftHeaderElement).toBeTruthy();
  });

  test.skip('must contain a right header element (drawer menu)', () => {
    const navigationOptions = IndexScreen.navigationOptions({ navigation });
    const rightHeaderElement = navigationOptions.headerRight;

    expect(rightHeaderElement).toBeTruthy();
  });
});
