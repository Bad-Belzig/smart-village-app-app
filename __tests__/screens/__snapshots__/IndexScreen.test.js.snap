// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndexScreen renders correctly 1`] = `null`;
